src/viewer.js

```diff
@@ -4008,49 +4012,93 @@ function doViewerResize(viewer, containerSize){
+    const pixels = viewer.pixelsPerArrowPress;
+    const panDelta = viewer.viewport.deltaPointsFromPixels(new OpenSeadragon.Point(pixels, pixels));
+    const panDelta40 = viewer.viewport.deltaPointsFromPixels(new OpenSeadragon.Point(40, 40));
+
     // Shift key state
     const shift = isDown('Shift') || isDown('ShiftLeft') || isDown('ShiftRight');
 
-    // Up Arrow
-    if (!viewer.preventVerticalPan && isDown('ArrowUp')) {
-        if (shift) {
-            viewer.viewport.zoomBy(1.1);
-        } else {
-            viewer.viewport.panBy(new OpenSeadragon.Point(0, -panDelta.y));
+    // Up Arrow and Down Arrow
+    if (!viewer.preventVerticalPan) {
+        if (isDown('ArrowUp')) {
+            if (shift) {
+                viewer.viewport.zoomBy(1.1);
+            } else {
+                viewer.viewport.panBy(new OpenSeadragon.Point(0, -panDelta.y));
+            }
+            viewer.viewport.applyConstraints();
+        }
+        if (isDown('ArrowDown')) {
+            if (shift) {
+                viewer.viewport.zoomBy(0.9);
+            } else {
+                viewer.viewport.panBy(new OpenSeadragon.Point(0, panDelta.y));
+            }
+            viewer.viewport.applyConstraints();
         }
-        viewer.viewport.applyConstraints();
     }
 
-    // Down Arrow
-    if (!viewer.preventVerticalPan && isDown('ArrowDown')) {
-        if (shift) {
-            viewer.viewport.zoomBy(0.9);
-        } else {
-            viewer.viewport.panBy(new OpenSeadragon.Point(0, panDelta.y));
+    // Left Arrow and Right Arrow
+    if (!viewer.preventHorizontalPan) {
+        if (isDown('ArrowLeft')) {
+            viewer.viewport.panBy(new OpenSeadragon.Point(-panDelta.x, 0));
+            viewer.viewport.applyConstraints();
+        }
+        if (isDown('ArrowRight')) {
+            viewer.viewport.panBy(new OpenSeadragon.Point(panDelta.x, 0));
+            viewer.viewport.applyConstraints();
         }
-        viewer.viewport.applyConstraints();
     }
 
-    // Left Arrow
-    if (isDown('ArrowLeft') && !viewer.preventHorizontalPan) {
-        viewer.viewport.panBy(new OpenSeadragon.Point(-panDelta.x, 0));
-        viewer.viewport.applyConstraints();
+    // WASD
+    if (!viewer.preventVerticalPan) {
+        if (isDown('KeyW')) {
+            if (shift) {
+                viewer.viewport.zoomBy(1.1);
+            } else {
+                viewer.viewport.panBy(new OpenSeadragon.Point(0, -panDelta40.y));
+            }
+            viewer.viewport.applyConstraints();
+        }
+        if (isDown('KeyS')) {
+            if (shift) {
+                viewer.viewport.zoomBy(0.9);
+            } else {
+                viewer.viewport.panBy(new OpenSeadragon.Point(0, panDelta40.y));
+            }
+            viewer.viewport.applyConstraints();
+        }
+    }
+
+    if (!viewer.preventHorizontalPan) {
+        if (isDown('KeyA')) {
+            viewer.viewport.panBy(new OpenSeadragon.Point(-panDelta40.x, 0));
```

Comment:
I see you have faithfully reproduced the 40 from the original code in `onCanvasKeyDown`. I believe that's actually a bug... the WASD keys and the arrow keys should behave the same. Let's use the regular `panDelta`.

Reply:
Once you've done that, the arrow and WASD keys should be identical, so you can combine their code.